Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift

```diff
@@ -194,3 +204,123 @@ private extension FetchResultSnapshotsProvider {
         }
     }
 }
+
+// MARK: - ObjectsDidChange Notification Handling
+
+@available(iOS 13.0, *)
+private extension FetchResultSnapshotsProvider {
+
+    /// The NotificationCenter to use for observing notifications.
+    var notificationCenter: NotificationCenter {
+        NotificationCenter.default
+    }
+
+    /// Start observing `NSManagedObjectContextObjectsDidChange` notifications so that snapshots
+    /// for object updates will be emitted.
+    ///
+    /// - SeeAlso: maybeEmitSnapshotFromObjectsDidChangeNotification
+    func startObservingObjectsDidChangeNotifications() {
+        // Remove token in case this method was called already.
+        stopObservingObjectsDidChangeNotifications()
+
+        objectsDidChangeObservationToken =
+            notificationCenter.addObserver(forName: .NSManagedObjectContextObjectsDidChange, object: storage, queue: nil) { [weak self] notification in
+                if let self = self {
+                    self.maybeEmitSnapshotFromObjectsDidChangeNotification(notification)
```

Comment:
[Q] When we end up with multiple `FetchResultSnapshotsProvider`s do you expect this function to become something that could block the UI while all of our instances process the change information?

Reply:
That's a good point. I tried to make `maybeEmitSnapshotFromObjectsDidChangeNotification` as performant as it can be. For example, I delayed the creation of `currentObjectIDs` since I wasn't sure of its performance. 

https://github.com/woocommerce/woocommerce-ios/blob/d7c214dcf188f034fe321b0376613edb70119a37/Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift#L297

The `maybeEmitSnapshotFromObjectsDidChangeNotification` method will get called whenever there's a `saveIfNeeded()` on a background (derived) storage. Usually, that's 25 rows (our default page size). Let's consider having 10 `SnapshotsProviders` open, which can be pretty high based on my understanding of the app. That's around `(10 * 25) + (10 * 25) + (10 * 25) + (10 * 25) = 1000` loops in the worst case. That's considering [`intersection`](https://developer.apple.com/documentation/swift/set/1779150-intersection) to be `O(n)` too (I'm not really sure what the performance of that is).  

I think 1000 loops on the main thread are negligible. 

I took a sample on an iPhone 8 Plus and here's what I got:

```
shiki: start maybeEmitSnapshotFromObjectsDidChangeNotification rows = 26
shiki: ended filter 0.0680685043334961 ms
shiki: ended objectIDsDiff 0.030994415283203125 ms
shiki: ended intersection 0.011920928955078125 ms
shiki: ended newSnapshot 0.026941299438476562 ms
shiki: ended sending new snapshot 27.009963989257812 ms
shiki: ended maybeEmitSnapshotFromObjectsDidChangeNotification 27.25505828857422 ms
```

For 26 updated rows, everything performed well except for this:

https://github.com/woocommerce/woocommerce-ios/blob/d7c214dcf188f034fe321b0376613edb70119a37/Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift#L310

That part took 27ms and is only because that call would cause the `UITableView` to update itself. It will block the main thread no matter what we do. 

Using that sample, the loops for 1000 rows are just `(27.25505828857422 - 27.009963989257812) * 5 = 1ms`. As for the `UITableView`, the `27ms` is probably close to the upper limit cause I think the cost of update is proportional to the number of visible rows. 

---

In summary, based on my limited and not science-backed performance testing, I think we're okay here. 😅 A future enhancement would be making everything run on the background thread. I learned that `dataSource.apply(snapshot)` can be called on a background thread so we can gain something for that. But I think it's probably over-engineering at this point. For now, I think it's best if we focus on fixing the crashes first.

Reply:
Thanks a lot for the analysis! 🙇